src/currentprocess.rs

```diff
@@ -139,20 +140,89 @@ pub fn with<F, R>(process: Process, f: F) -> R
 where
     F: FnOnce() -> R,
 {
+    ensure_hook();
+
+    PROCESS.with(|p| {
+        if let Some(old_p) = &*p.borrow() {
+            panic!("current process already set {old_p:?}");
+        }
+        *p.borrow_mut() = Some(process);
+        let result = f();
+        *p.borrow_mut() = None;
+        result
+    })
+}
+
+fn ensure_hook() {
     HOOK_INSTALLED.call_once(|| {
         let orig_hook = panic::take_hook();
         panic::set_hook(Box::new(move |info| {
             clear_process();
             orig_hook(info);
         }));
     });
+}
 
-    PROCESS.with(|p| {
+/// Run a function in the context of a process definition and a tokio runtime.
+///
+/// The process state is injected into a thread-local in every work thread of
+/// the runtime, but this requires access to the runtime builder, so this
+/// function must be the one to create the runtime.
+pub fn with_runtime<'a, FR, R>(
+    process: Process,
+    mut runtime_builder: tokio::runtime::Builder,
+    fut: FR,
```

Comment:
Suggest writing this as `impl Future<Output = R> + 'a`.

Reply:
done